src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs

```diff
@@ -193,6 +201,111 @@ private List<string> ResolveByExternalReferencePackages(List<string> uidList, Co
             return list;
         }
 
+        private async Task<List<string>> ResolveByXRefServiceAsync(List<string> uidList, ConcurrentDictionary<string, XRefSpec> externalXRefSpec)
+        {
+            if (XRefServiceUrls == null || XRefServiceUrls.Length == 0)
+            {
+                return uidList;
+            }
+            
+            var unresolvedUidList = new List<string>();
+            int pieceSize = 1000;
+            for(int i = 0; i < uidList.Count; i += pieceSize)
+            {
+                List<string> smallPiece = uidList.GetRange(i, Math.Min(pieceSize, uidList.Count - i));
+                foreach(string requestUrl in XRefServiceUrls)
+                {
+                    var queryUidTasks = new List<Task<List<XRefSpec>>>();
+                    foreach(string uid in smallPiece)
+                    {
+                        queryUidTasks.Add(QueryByHttpRequestAsync(requestUrl, uid));
+                    }
+                    var results = await Task.WhenAll(queryUidTasks);
+                    List<string> stillNeedQueryList = new List<string>();
+                    for(int j = 0; j < results.Length; j++)
+                    {
+                        if(results[j].Count > 0 && results[j][0].Uid == smallPiece[j])
+                        {
+                            externalXRefSpec.AddOrUpdate(smallPiece[j], results[j][0], (_, old) => old + results[j][0]);
+                        }
+                        else
+                        {
+                            stillNeedQueryList.Add(smallPiece[j]);
+                        }
+                    }
+                    smallPiece = stillNeedQueryList;
+                }
+                unresolvedUidList.AddRange(smallPiece);
+            }
+
+            Logger.LogInfo($"{uidList.Count - unresolvedUidList.Count} uids found in xrefservice.");
+            return unresolvedUidList;
+        }
+
+        public async Task<List<XRefSpec>> QueryByHttpRequestAsync(string requestUrl, string uid)
+        {
+            using (var client = new HttpClient())
```

Comment:
No need to create instance per request.
See [https://msdn.microsoft.com/en-us/library/system.net.http.httpclient(v=vs.110).aspx#Remarks](https://msdn.microsoft.com/en-us/library/system.net.http.httpclient(v=vs.110).aspx#Remarks)
GetAsync is thread safety.